x-pack/plugins/lens/public/data_views_service/service.ts

```diff
@@ -122,7 +124,12 @@ export function createIndexPatternService({
     },
     replaceDataViewId: async (dataView: DataView) => {
       const newDataView = await dataViews.create({ ...dataView.toSpec(), id: generateId() });
-      dataViews.clearInstanceCache(dataView.id);
+
+      // Do not clear initial data view instance from cache
+      // if adhoc data view id has been provided by the context.
+      if (contextDataViewSpec && contextDataViewSpec.id !== dataView.id) {
+        dataViews.clearInstanceCache(dataView.id);
```

Comment:
Why do we need to keep the old data view in cache here? When navigating back to discover the spec is passed over so it should be able to re-create from that.

Reply:
`spec` isn't really passed over, since Discover -> Lens navigation uses `ui-actions`. Going back will  route to URL  state with data view id.

Reply:
You mean going back via browser back button?

Reply:
Yes

Reply:
Hmm, alright, let's keep it in like this. It's not 100% reliable as the user could reload the page while in Lens, then navigate back which would break Discover. Seems like an edge case though. cc @kertal maybe we can find a more stable way to handle this longterm.

Reply:
yes, if we find a better way here to prevent user running into ad-hoc data views that for various reasons no longer exist, it would be awesome, also with increased adaptations by our userbase I'm pretty sure this will be requested